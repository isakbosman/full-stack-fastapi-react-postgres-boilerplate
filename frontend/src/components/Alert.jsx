import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import theme from 'modules/theme';

import { Alert as AlertComponent, Box, utils } from 'styled-minimal';
import Icon from 'components/Icon';

const { colors, palette } = utils.getTheme(theme);
const variants = { ...colors, ...palette };

AlertComponent.displayName = 'AlertComponent';

const AlertWrapper = styled(AlertComponent)`
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
`;

const AlertIcon = styled.div`
  align-items: flex-start;
  background-color: ${({ variant }) => variants[variant]};
  color: #fff;
  display: flex;
  padding: ${utils.spacer(3)};
`;

const AlertButton = styled.button`
  background-color: ${({ variant }) => variants[variant]};
  color: #ccc;
  pointer-events: all;
  position: absolute;
  right: ${utils.spacer(1)};
  top: ${utils.spacer(1)};
`;

const Alert = ({ children, handleClickClose, id, icon, ...rest }) => {
  const output = {};
  let name;

  switch (rest.variant) {
    case 'success': {
      name = icon || 'check-circle';
      break;
    }
    case 'warning': {
      name = icon || 'exclamation-circle';
      break;
    }
    case 'danger': {
      name = icon || 'times-circle';
      break;
    }
    case 'info': {
      name = icon || 'question-circle';
      break;
    }
    case 'dark': {
      name = icon || 'bell-o';
      break;
    }
    default: {
      name = icon || 'dot-circle-o';
    }
  }

  if (handleClickClose) {
    output.button = (
      <AlertButton data-id={id} onClick={handleClickClose} type="button" data-testid="AlertButton">
        <Icon name="times" width={10} />
      </AlertButton>
    );
  }

  return (
    <AlertWrapper {...rest} data-testid="AlertWrapper">
      <AlertIcon {...rest}>
        <Icon name={name} width={24} />
      </AlertIcon>
      <Box p={2} pr={handleClickClose ? 3 : 2}>
        {children}
      </Box>
      {output.button}
    </AlertWrapper>
  );
};

Alert.propTypes = {
  children: PropTypes.node.isRequired,
  handleClickClose: PropTypes.func,
  icon: PropTypes.string,
  id: PropTypes.string,
  outline: PropTypes.bool,
  variant: PropTypes.string,
};

Alert.defaultProps = {
  outline: true,
  variant: 'info',
};

export default Alert;
