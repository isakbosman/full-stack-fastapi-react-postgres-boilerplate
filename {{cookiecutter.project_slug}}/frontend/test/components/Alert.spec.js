import React from 'react';

import Alert from 'components/Alert';

function setup(children = 'Hello World', variant) {
  return mount(<Alert variant={variant}>{children}</Alert>);
}

describe('Alert', () => {
  let wrapper = setup();

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a success alert', () => {
    wrapper = setup('This is a success message', 'success');

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a error alert with markup', () => {
    wrapper = setup(<p>This is an error message</p>, 'danger');

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a warning alert', () => {
    wrapper = setup('This is a warning message', 'warning');

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a info alert', () => {
    wrapper = setup('This is an info message', 'info');

    expect(wrapper).toMatchSnapshot();
  });
});
