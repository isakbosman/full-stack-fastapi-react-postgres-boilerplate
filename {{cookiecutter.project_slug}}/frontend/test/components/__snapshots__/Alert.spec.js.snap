// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert should render a error alert with markup 1`] = `
.c2 {
  box-sizing: border-box;
  padding: 8px;
  padding-right: 8px;
}

.c0 {
  box-sizing: border-box;
  background-color: #ff4444;
  border: 1px solid #ff4444;
  color: #fff;
  border-radius: 3px;
  font-size: 16px;
  line-height: 1.4;
  max-width: 450px;
  padding: 10px 16px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
}

.c0 a {
  color: #ff4444;
}

.c1 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: #ff4444;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px;
}

<AlertWrapper
  bordered={false}
  dark={false}
  data-testid="AlertWrapper"
  outline={true}
  size="md"
  variant="danger"
>
  <div
    className="c0"
    data-testid="AlertWrapper"
    role="alert"
    size="md"
  >
    <AlertIcon
      outline={true}
      variant="danger"
    >
      <div
        className="c1"
      >
        <IconWrapper
          src="/media/icons/times-circle.svg"
          width={24}
        />
      </div>
    </AlertIcon>
    <Box
      p={2}
      pr={2}
    >
      <div
        className="c2"
      >
        <p>
          This is an error message
        </p>
      </div>
    </Box>
  </div>
</AlertWrapper>
`;

exports[`Alert should render a info alert 1`] = `
.c2 {
  box-sizing: border-box;
  padding: 8px;
  padding-right: 8px;
}

.c0 {
  box-sizing: border-box;
  background-color: #33b5e5;
  border: 1px solid #33b5e5;
  color: #fff;
  border-radius: 3px;
  font-size: 16px;
  line-height: 1.4;
  max-width: 450px;
  padding: 10px 16px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
}

.c0 a {
  color: #33b5e5;
}

.c1 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: #33b5e5;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px;
}

<AlertWrapper
  bordered={false}
  dark={false}
  data-testid="AlertWrapper"
  outline={true}
  size="md"
  variant="info"
>
  <div
    className="c0"
    data-testid="AlertWrapper"
    role="alert"
    size="md"
  >
    <AlertIcon
      outline={true}
      variant="info"
    >
      <div
        className="c1"
      >
        <IconWrapper
          src="/media/icons/question-circle.svg"
          width={24}
        />
      </div>
    </AlertIcon>
    <Box
      p={2}
      pr={2}
    >
      <div
        className="c2"
      >
        This is an info message
      </div>
    </Box>
  </div>
</AlertWrapper>
`;

exports[`Alert should render a success alert 1`] = `
.c2 {
  box-sizing: border-box;
  padding: 8px;
  padding-right: 8px;
}

.c0 {
  box-sizing: border-box;
  background-color: #00C851;
  border: 1px solid #00C851;
  color: #fff;
  border-radius: 3px;
  font-size: 16px;
  line-height: 1.4;
  max-width: 450px;
  padding: 10px 16px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
}

.c0 a {
  color: #00C851;
}

.c1 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: #00C851;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px;
}

<AlertWrapper
  bordered={false}
  dark={false}
  data-testid="AlertWrapper"
  outline={true}
  size="md"
  variant="success"
>
  <div
    className="c0"
    data-testid="AlertWrapper"
    role="alert"
    size="md"
  >
    <AlertIcon
      outline={true}
      variant="success"
    >
      <div
        className="c1"
      >
        <IconWrapper
          src="/media/icons/check-circle.svg"
          width={24}
        />
      </div>
    </AlertIcon>
    <Box
      p={2}
      pr={2}
    >
      <div
        className="c2"
      >
        This is a success message
      </div>
    </Box>
  </div>
</AlertWrapper>
`;

exports[`Alert should render a warning alert 1`] = `
.c2 {
  box-sizing: border-box;
  padding: 8px;
  padding-right: 8px;
}

.c0 {
  box-sizing: border-box;
  background-color: #fda509;
  border: 1px solid #fda509;
  color: #fff;
  border-radius: 3px;
  font-size: 16px;
  line-height: 1.4;
  max-width: 450px;
  padding: 10px 16px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
}

.c0 a {
  color: #fda509;
}

.c1 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: #fda509;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px;
}

<AlertWrapper
  bordered={false}
  dark={false}
  data-testid="AlertWrapper"
  outline={true}
  size="md"
  variant="warning"
>
  <div
    className="c0"
    data-testid="AlertWrapper"
    role="alert"
    size="md"
  >
    <AlertIcon
      outline={true}
      variant="warning"
    >
      <div
        className="c1"
      >
        <IconWrapper
          src="/media/icons/exclamation-circle.svg"
          width={24}
        />
      </div>
    </AlertIcon>
    <Box
      p={2}
      pr={2}
    >
      <div
        className="c2"
      >
        This is a warning message
      </div>
    </Box>
  </div>
</AlertWrapper>
`;

exports[`Alert should render properly 1`] = `
.c2 {
  box-sizing: border-box;
  padding: 8px;
  padding-right: 8px;
}

.c0 {
  box-sizing: border-box;
  background-color: #33b5e5;
  border: 1px solid #33b5e5;
  color: #fff;
  border-radius: 3px;
  font-size: 16px;
  line-height: 1.4;
  max-width: 450px;
  padding: 10px 16px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1;
  padding: 0;
  position: relative;
}

.c0 a {
  color: #33b5e5;
}

.c1 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: #33b5e5;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px;
}

<AlertWrapper
  bordered={false}
  dark={false}
  data-testid="AlertWrapper"
  outline={true}
  size="md"
  variant="info"
>
  <div
    className="c0"
    data-testid="AlertWrapper"
    role="alert"
    size="md"
  >
    <AlertIcon
      outline={true}
      variant="info"
    >
      <div
        className="c1"
      >
        <IconWrapper
          src="/media/icons/question-circle.svg"
          width={24}
        />
      </div>
    </AlertIcon>
    <Box
      p={2}
      pr={2}
    >
      <div
        className="c2"
      >
        Hello World
      </div>
    </Box>
  </div>
</AlertWrapper>
`;
